File: FeynmanDiagram/phononMethods.py
import sys

sys.path.append('../')
from Propagator import D

def addInternalPhonon(self, v1, v2, momentum, theta, phi):
  d = D(momentum)

  if v1.position > v2.position:
    print('ERROR: Phonon cannot run backwards')
    return False

  d.setStart(v1)
  d.setEnd(v2)

  # append to list
  self.Ds.append(d)

  # change momentum for electrons under phonon for conservation of momentum
  g = v1.G[1]
  while g.start != v2:
    g.addMomentum(-momentum)
    g = g.end.G[1]

  ##
  ## NEED TO ALSO GIVE THE PHONON ITS ANGLE
  ##
  self.setInternalPhononMomentumAngle(d, theta, phi)

  # return phonon
  return d

def removeInternalPhonon(self, dIndex):
  d = self.Ds[dIndex]

  # change momentum for electrons under phonon for conservation of momentum
  g = d.start.G[1]
  while g.start != d.end:
    g.addMomentum(d.momentum)
    g = g.end.G[1]

  # will unlink from vertices
  d.remove()

  # remove from Ds list                                                            <-- takes time => hash table
  del self.Ds[dIndex]

def setInternalPhononMomentum(self, d, momentum):
  dP = momentum - d.momentum

  # add momentum difference
  d.addMomentum(dP)

  # change momentum for electrons under phonon for conservation of momentum
  g = d.start.G[1]
  while g.start != d.end:
    g.addMomentum(-dP)
    g = g.end.G[1]

  # return changed phonon
  return d

def setInternalPhononMomentumAngle(self, d, theta, phi):
  # here we assume that the momentum has already been updated by the method above
  d.setTheta(theta)
  d.setPhi(phi)

##
## take a look at this when simply reversing phonon
##
def swapPhononEnds(self, v1, v2):
  # order temporally
  if v1.position > v2.position:
    (v1, v2) = (v2, v1)

  # need to be neighbours
  if v1.G[1].end != v2:
    print('ERROR: vertices not neighbours')
    return False
    
  d1 = v1.D[0] or v1.D[1]
  d2 = v2.D[0] or v2.D[1]

  # change in momentum
  dp1 = -d1.momentum if v1.D[0] else d1.momentum
  dp2 = -d2.momentum if v2.D[1] else d2.momentum
  dp = dp1 + dp2

  # conserve momentum
  v1.G[1].addMomentum(dp)

  # relink
  if d1 == d2:
    # just reverse the momentum
    d1.setMomentum(-d1.momentum)
  else:
    if d1.start == v1:
      d1.setStart(v2)
    else:
      d1.setEnd(v2)

    if d2.start == v2:
      d2.setStart(v1)
    else:
      d2.setEnd(v1)

File: src/FeynmanDiagram/phononMethods.cpp
#include "FeynmanDiagram.h"

shared_ptr<Phonon> FeynmanDiagram::addInternalPhonon (
  shared_ptr<Vertex> v1,
  shared_ptr<Vertex> v2,
  Vector3d Q, 
  double q,
  double theta,
  double phi
) {
  // first make sure that v1 and v2 are not already attached to a phonon
  if (v1->D[0] != NULL or v1->D[1] != NULL or v2->D[0] != NULL or v2->D[1] != NULL) {
    cout << "ERROR at FeynmanDiagram::addInternalPhonon: a phonon is already attached to either/both of these vertices" << endl;
    return NULL;
  }

  // create new phonon
  this->Ds.emplace_back(new Phonon(Q, q, theta, phi));

  // attach to vertices
  this->Ds.back()->setStart(v1);
  this->Ds.back()->setEnd(v2);

  // momentum conservation
  shared_ptr<Electron> g = v1->G[1];
  do {
    // remove from hash table
    this->removeFromHashTable(g);  

    g->addMomentum(-Q);

    // reinsert into hash table
    this->insertIntoHashTable(g);
  } while (g->end != v2 && (g = g->end->G[1]));

  return this->Ds.back();
}

void FeynmanDiagram::removeInternalPhonon (unsigned int phononIndex) {
  shared_ptr<Phonon> d = this->Ds[phononIndex];

  // momentum conservation
  shared_ptr<Electron> g = d->start->G[1];

  do {
    // remove from hash table
    this->removeFromHashTable(g);  

    g->addMomentum(d->momentum);

    // reinsert into hash table
    this->insertIntoHashTable(g);
  } while (g->end != d->end && (g = g->end->G[1]));

  // unlink phonon from vertices
  d->setStart(NULL);
  d->setEnd(NULL);

  // remove d from Ds
  this->Ds.erase(this->Ds.begin() + phononIndex);
}

void FeynmanDiagram::setInternalPhononMomentum (shared_ptr<Phonon> d, Vector3d Q, double q) {
  // obviously need to be calculated before adding the momentum
  Vector3d dQ = Q - d->momentum;

  // add momentum difference
  d->setMomentum(Q, q);

  // momentum conservation
  shared_ptr<Electron> g = d->start->G[1];
  do {
    // remove from hash table
    this->removeFromHashTable(g);  

    g->addMomentum(-dQ);
    
    // reinsert into hash table
    this->insertIntoHashTable(g);
  } while (g->end != d->end && (g = g->end->G[1]));
}

void FeynmanDiagram::setInternalPhononMomentumDirection (shared_ptr<Phonon> d, double theta, double phi) {
  d->setTheta(theta);
  d->setPhi(phi);
}